subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformValuesInjectionIntegrationTest.groovy

```diff
@@ -547,6 +548,51 @@ class ArtifactTransformValuesInjectionIntegrationTest extends AbstractDependency
         ann << [OutputFile, OutputFiles, OutputDirectory, OutputDirectories, Destroys, LocalState, OptionValues]
     }
 
+    @ValidationTestFor(
+        ValidationProblemId.ANNOTATION_INVALID_IN_CONTEXT
+    )
+    @Unroll
+    def "transform parameters type cannot use annotation @Untracked"() {
+        settingsFile << """
+            include 'a', 'b'
+        """
+        setupBuildWithColorTransform()
+        buildFile << """
+            project(':a') {
+                dependencies {
+                    implementation project(':b')
+                }
+            }
+
+            abstract class MakeGreen implements TransformAction<Parameters> {
+                interface Parameters extends TransformParameters {
+                    @Untracked
+                    @InputFiles
+                    File getBad()
+                    void setBad(File value)
+                }
+
+                void transform(TransformOutputs outputs) {
+                    throw new RuntimeException()
+                }
+            }
+"""
+
+        when:
+        fails(":a:resolve")
+
+        then:
+        failure.assertHasDescription("Execution failed for task ':a:resolve'.")
+        failure.assertHasCause("Could not resolve all files for configuration ':a:implementation'.")
+        failure.assertHasCause("Failed to transform b.jar (project :b) to match attributes {artifactType=jar, color=green}.")
+        failure.assertThatCause(matchesRegexp('Could not isolate parameters MakeGreen\\$Parameters_Decorated@.* of artifact transform MakeGreen'))
+        failure.assertHasCause('A problem was found with the configuration of the artifact transform parameter MakeGreen.Parameters.')
+        assertPropertyValidationErrors(bad: modifierAnnotationInvalidInContext {
+            annotation(Untracked.simpleName)
+            validAnnotations = "@Classpath, @CompileClasspath, @IgnoreEmptyDirectories, @Incremental, @NormalizeLineEndings, @Optional or @PathSensitive"
```

Comment:
These should be a list of actual type references if we are serious about this fluent API approach, no? Or, if it's okay to use a string here, why aren't we testing the whole message as a string? Why draw the line right here?

Reply:
Like said above, let's have this discussion separately. For this particular case, I think a few things come together: We only have one case where we currently use the new `modifierAnnotationInvalidInContext()`. So I did not add a new delegate class for it, but used the same as for `annotationInvalidInContext`. In a similar vein I didn't add a new method to set these `validAnnotations` in a fluent way. If we could declare those fixtures in shared places, and wouldn't be forced to do all of them in one place, then I probably would have chosen a different approach. I think we only need very good/fluent fixtures for stuff we use many times, and don't have to be so strict for things which we only use once or twice.

Still, let's have this discussion separately.